### PR TITLE
Implemented in IR the managed methods of the Saturating region ...

### PR DESCRIPTION
**Pull request by LDj3SNuD**
... of the SoftFallback class (the SatQ ones).

The need to natively manage the Fpcr and Fpsr system registers is still a fact.

Contributes to https://github.com/Ryujinx/Ryujinx/issues/2917 ; I will open another PR to implement in Intrinsics-branchless the methods of the Saturation region as well (the SatXXXToXXX ones).

All instructions involved have been tested locally in both release and debug modes, in both lowcq and highcq.

Jit disasm of some instructions involved:
[0x000002216B46102C_297_Suqadd_S_D.txt](https://github.com/Ryujinx/Ryujinx/files/9479703/0x000002216B46102C_297_Suqadd_S_D.txt)
[0x00000189803B102C_254_Usqadd_S_D.txt](https://github.com/Ryujinx/Ryujinx/files/9479704/0x00000189803B102C_254_Usqadd_S_D.txt)
[0x000001AE5F39102C_248_Sqadd_S_D.txt](https://github.com/Ryujinx/Ryujinx/files/9479705/0x000001AE5F39102C_248_Sqadd_S_D.txt)
[0x00000274A693102C_199_Uqadd_S_D.txt](https://github.com/Ryujinx/Ryujinx/files/9479706/0x00000274A693102C_199_Uqadd_S_D.txt)
[0x0000014EF028102C_193_Sqabs_S_D.txt](https://github.com/Ryujinx/Ryujinx/files/9479707/0x0000014EF028102C_193_Sqabs_S_D.txt)
[0x000001FF7824102C_174_Uqxtn_S_DS.txt](https://github.com/Ryujinx/Ryujinx/files/9479708/0x000001FF7824102C_174_Uqxtn_S_DS.txt)
[0x000001FF7824102C_195_Sqxtn_S_DS.txt](https://github.com/Ryujinx/Ryujinx/files/9479709/0x000001FF7824102C_195_Sqxtn_S_DS.txt)

### CODE REVIEW

**Comment by gdkchan** [comment]
Faster video decoding on Tony Hawk's Pro Skater intro:
master:

https://user-images.githubusercontent.com/5624669/188240450-19793a0a-fa66-44bf-94b8-385d36065e67.mp4

PR:

https://user-images.githubusercontent.com/5624669/188240724-69f78e9a-1916-4895-ab30-77c6abb3dc5f.mp4

Higher framerate inside the HQ on Astral Chain:
master:

image

PR:

image

Pre-rendered video decoding is also much faster on this game with those changes.

**Review by gdkchan — APPROVED** [review]
Lgtm, it would be better if we didn't need to call managed to set the QC flag. Nice work optimizing this!

**Comment by EmulationFanatic** [comment]
Faster video decoding on Crash Team Nitro Racing.
Master:
2022-09-02 17_12_46-Ryujinx 1 1 248 - Crash™ Team Racing Nitro-Fueled v1 0 15 (0100F9F00C696000) (64

PR:
2022-09-02 16_54_41-Ryujinx 1 1 0+8b1430f - Crash™ Team Racing Nitro-Fueled v1 0 15 (0100F9F00C69600



Faster (but still not perfect) video decoding with Mario + Rabbids Kingdom Battle:
Master:
2022-09-02 16_46_54-Ryujinx 1 1 248 - Mario + Rabbids Kingdom Battle v1 0 371506 (010067300059A000)

PR:

2022-09-02 16_48_33-Ryujinx 1 1 0+8b1430f - Mario + Rabbids Kingdom Battle v1 0 371506 (010067300059

File: ARMeilleure/Instructions/InstEmitSimdHelper.cs [hunk 1]
```diff
@@ -1520,82 +1523,260 @@ public static void EmitSaturatingNarrowOp(ArmEmitterContext context, SaturatingN
...
+
+            Operand maxUL = Const(ulong.MaxValue);
+
+            Operand res = context.Copy(context.AllocateLocal(OperandType.I64), context.Add(op1, op2));
 
-            MethodInfo info = signed
-                ? typeof(SoftFallback).GetMethod(nameof(SoftFallback.BinarySignedSatQAcc))
-                : typeof(SoftFallback).GetMethod(nameof(SoftFallback.BinaryUnsignedSatQAcc));
+            context.BranchIf(lblEnd, res, op1, Comparison.GreaterOrEqualUI);
+            context.BranchIf(lblEnd, res, op2, Comparison.GreaterOrEqualUI);
```
Comment (merryhime) [inline ARMeilleure/Instructions/InstEmitSimdHelper.cs:1653]: This is reducible to a single branch. If result is `u<` op1, an overflow has occurred.

**Comment by gdkchan** [comment]
`SetFpsrQc` is being called a lot on Tony Hawk's Pro Skater. I beleive that also moving that to the native side is required to further improve performance here.

**Comment by MetrosexualGarbodor** [comment]
The difference in the Astral Chain intro is really big: ~23fps on master and around 80-100fps on this PR, running Vulkan on both.

image

image


In the police HQ, it's 29-30 on master and 42-43 on this PR.

image

image


Very nice work.

**Review by riperiperi — APPROVED** [review]
LGTM, nice to see these softfallback calls removed. Only calling into managed to set FPSR when saturation happens is much better, but it would be nice to have the flag entirely settable from native in future.

**Comment by gdkchan** [comment]
I know you're working on further improvements here. Do you plan to include them as part of this PR or a different one? If you plan to include as part of this one, I'd convert it to draft until its ready.

**Comment by LDj3SNuD** [comment]
> I know you're working on further improvements here. Do you plan to include them as part of this PR or a different one? If you plan to include as part of this one, I'd convert it to draft until its ready.

image
different one (shlreg)

**Comment by FICTURE7** [comment]
A less intrusive way to get rid of the `SetFpsrQc` call other than modelling the flags using `FpFlag` is to move `Fpsr` in the `ExecutionContext` to the `NativeContext.NativeStorageCtx`, then accessing them through loads/stores in IR. Something similar can be done for `Fpcr`. Of course, modelling the flags using `FpFlag` (or `Flag`?) is probably the most adequate solution.

```patch
diff --git a/ARMeilleure/Instructions/InstEmitHelper.cs b/ARMeilleure/Instructions/InstEmitHelper.cs
index a22bb3fb..ea561681 100644
--- a/ARMeilleure/Instructions/InstEmitHelper.cs
+++ b/ARMeilleure/Instructions/InstEmitHelper.cs
@@ -260,5 +260,13 @@ namespace ARMeilleure.Instructions
         {
             context.Copy(GetFpFlag(stateFlag), value);
         }
+
+        public static void SetFpsrQc(ArmEmitterContext context)
+        {
+            Operand nativeContext = context.LoadArgument(OperandType.I64, 0);
+            Operand fpsrAddr = context.Add(nativeContext, Const((long)NativeContext.GetFpsrOffset()));
+            Operand fpsr = context.Load(OperandType.I32, fpsrAddr);
+            context.Store(fpsrAddr, context.BitwiseOr(fpsr, Const((int)FPSR.Qc)));
+        }
     }
 }
diff --git a/ARMeilleure/Instructions/InstEmitSimdShift32.cs b/ARMeilleure/Instructions/InstEmitSimdShift32.cs
index 6dcfe065..d7d03f35 100644
--- a/ARMeilleure/Instructions/InstEmitSimdShift32.cs
+++ b/ARMeilleure/Instructions/InstEmitSimdShift32.cs
@@ -1,5 +1,6 @@
 ﻿using ARMeilleure.Decoders;
 using ARMeilleure.IntermediateRepresentation;
+using ARMeilleure.State;
 using ARMeilleure.Translation;
 using System;
 using System.Diagnostics;
@@ -330,7 +331,7 @@ namespace ARMeilleure.Instructions
 
             context.BranchIfFalse(lblNoSat, context.BitwiseOr(gt, lt));
 
-            context.Call(typeof(NativeInterface).GetMethod(nameof(NativeInterface.SetFpsrQc)));
+            InstEmitHelper.SetFpsrQc(context);
 
             context.MarkLabel(lblNoSat);
 
diff --git a/ARMeilleure/State/ExecutionContext.cs b/ARMeilleure/State/ExecutionContext.cs
index c73ca197..bd638b65 100644
--- a/ARMeilleure/State/ExecutionContext.cs
+++ b/ARMeilleure/State/ExecutionContext.cs
@@ -36,8 +36,18 @@ namespace ARMeilleure.State
             set => _nativeContext.SetPstate(value);
         }
 
-        public FPCR Fpcr { get; set; }
-        public FPSR Fpsr { get; set; }
+        public FPCR Fpcr
+        {
+            get => (FPCR)_nativeContext.GetFpcr();
+            set => _nativeContext.SetFpcr((uint)value);
+        }
+
+        public FPSR Fpsr
+        {
+            get => (FPSR)_nativeContext.GetFpsr();
+            set => _nativeContext.SetFpsr((uint)value);
+        }
+
         public FPCR StandardFpcrValue => (Fpcr & (FPCR.Ahp)) | FPCR.Dn | FPCR.Fz;
 
         public bool IsAarch32 { get; set; }
diff --git a/ARMeilleure/State/NativeContext.cs b/ARMeilleure/State/NativeContext.cs
index 11ab5ca3..2d8d7583 100644
--- a/ARMeilleure/State/NativeContext.cs
+++ b/ARMeilleure/State/NativeContext.cs
@@ -13,6 +13,8 @@ namespace ARMeilleure.State
             public fixed ulong V[RegisterConsts.VecRegsCount * 2];
             public fixed uint Flags[RegisterConsts.FlagsCount];
             public fixed uint FpFlags[RegisterConsts.FpFlagsCount];
+            public uint Fpcr;
+            public uint Fpsr;
             public int Counter;
             public ulong DispatchAddress;
             public ulong ExclusiveAddress;
@@ -120,6 +122,26 @@ namespace ARMeilleure.State
             }
         }
 
+        public uint GetFpcr()
+        {
+            return GetStorage().Fpcr;
+        }
+
+        public void SetFpcr(uint value)
+        {
+            GetStorage().Fpcr = value;
+        }
+
+        public uint GetFpsr()
+        {
+            return GetStorage().Fpcr;
+        }
+
+        public void SetFpsr(uint value)
+        {
+            GetStorage().Fpsr = value;
+        }
+
         public unsafe bool GetFPStateFlag(FPState flag)
         {
             if ((uint)flag >= RegisterConsts.FpFlagsCount)
@@ -186,6 +208,16 @@ namespace ARMeilleure.State
             }
         }
 
+        public static int GetFpcrOffset()
+        {
+            return StorageOffset(ref _dummyStorage, ref _dummyStorage.Fpcr);
+        }
+
+        public static int GetFpsrOffset()
+        {
+            return StorageOffset(ref _dummyStorage, ref _dummyStorage.Fpsr);
+        }
+
         public static int GetCounterOffset()
         {
             return StorageOffset(ref _dummyStorage, ref _dummyStorage.Counter);
```

**Pull request merged**